# 第六次作业

在本次作业中，我们利用R-GCN来完整实体分类任务。实体分类和节点分类任务相似，都是对图里的节点进行分类。

## 数据集加载

我们使用AIFB数据集来完成实体分类。R-GCN的论文里面就使用了这个数据集来完成实体分类任务。

In [1]:
import torch
import torch.nn.functional as F
from torch_geometric.datasets import Entities


dataset = Entities('./data', 'AIFB')
data = dataset[0]

node_idx = torch.cat([data.train_idx, data.test_idx], dim=0)

In [2]:
data

Data(edge_index=[2, 58086], edge_type=[58086], train_idx=[140], train_y=[140], test_idx=[36], test_y=[36], num_nodes=8285)

In [3]:
data.edge_type.max().item()+1

90

我们可以看到在AIFB数据集中，有58086条边，90种关系，8285个节点。其中有140个节点是训练集中的节点，36个节点是测试集中的节点。

## 代码填空
完成空缺的代码部分，完成实体分类任务。

In [4]:
# 大家既可以自己写RGCNConv，也可以用如下的代码调用RGCNConv
# from torch_geometric.nn import RGCNConv

class RGCN(torch.nn.Module):
    
    def __init__(self, num_nodes, hidden_channels, num_classes, num_relations):
        super().__init__()
        
        ###################
        #### 代码填空 ######
        ###################
        
    def reset_parameters(self):
        """初始化模型参数"""
        ###################
        #### 代码填空 ######
        ###################
    
    def forward(self, edge_index, edge_type):
        """前向传播"""
        ###################
        #### 代码填空 ######
        ###################

In [5]:
def train():
    """训练模型"""
    ###################
    #### 代码填空 ######
    ###################
    return loss.item()

In [6]:
@torch.no_grad()
def test():
    model.eval()
    pred = model(data.edge_index, data.edge_type).argmax(dim=-1)
    ###########################################
    #### 代码填空，计算train_acc和test_acc ######
    ##########################################     
    return train_acc.item(), test_acc.item()

In [7]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = RGCN(data.num_nodes, 16, dataset.num_classes, dataset.num_relations).to(device)
data = data.to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.01, weight_decay=0.0005)


for epoch in range(1, 100):
    loss = train()
    train_acc, test_acc = test()
    if epoch % 5 == 0:
        print(f'Epoch: {epoch:02d}, TrainLoss: {loss:.4f}, TrainAcc: {train_acc:.4f} '
              f'TestAcc: {test_acc:.4f}')

Epoch: 05, TrainLoss: 0.3687, TrainAcc: 0.9714 TestAcc: 0.8333
Epoch: 10, TrainLoss: 0.0234, TrainAcc: 1.0000 TestAcc: 0.9444
Epoch: 15, TrainLoss: 0.0015, TrainAcc: 1.0000 TestAcc: 0.9167
Epoch: 20, TrainLoss: 0.0002, TrainAcc: 1.0000 TestAcc: 0.8889
Epoch: 25, TrainLoss: 0.0001, TrainAcc: 1.0000 TestAcc: 0.8889
Epoch: 30, TrainLoss: 0.0000, TrainAcc: 1.0000 TestAcc: 0.8889
Epoch: 35, TrainLoss: 0.0001, TrainAcc: 1.0000 TestAcc: 0.8889
Epoch: 40, TrainLoss: 0.0001, TrainAcc: 1.0000 TestAcc: 0.8889
Epoch: 45, TrainLoss: 0.0004, TrainAcc: 1.0000 TestAcc: 0.9167
Epoch: 50, TrainLoss: 0.0009, TrainAcc: 1.0000 TestAcc: 0.9444
Epoch: 55, TrainLoss: 0.0016, TrainAcc: 1.0000 TestAcc: 0.9444
Epoch: 60, TrainLoss: 0.0023, TrainAcc: 1.0000 TestAcc: 0.9444
Epoch: 65, TrainLoss: 0.0026, TrainAcc: 1.0000 TestAcc: 0.9444
Epoch: 70, TrainLoss: 0.0026, TrainAcc: 1.0000 TestAcc: 0.9444
Epoch: 75, TrainLoss: 0.0022, TrainAcc: 1.0000 TestAcc: 0.9444
Epoch: 80, TrainLoss: 0.0020, TrainAcc: 1.0000 TestAcc: